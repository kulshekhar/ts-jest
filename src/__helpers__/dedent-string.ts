export const omitLeadingWhitespace = (text: string): string => {
  return text.replace(/^\s+/gm, '')
}

export const dedent = (strings: TemplateStringsArray, ...values: unknown[]) => {
  let joinedString = ''
  for (let i = 0; i < values.length; i++) {
    joinedString += `${strings[i]}${values[i]}`
  }
  joinedString += strings[strings.length - 1]

  return omitLeadingWhitespace(joinedString)
}
