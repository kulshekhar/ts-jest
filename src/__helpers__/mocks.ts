import { testing } from 'bs-logger'

import { rootLogger } from '../util/logger'

export const logTargetMock = () => (rootLogger as testing.LoggerMock).target

export const mockObject = <T, M>(obj: T, newProps: M): T & M & { mockRestore: () => T } => {
  const backup: any = Object.create(null)

  Object.keys(newProps).forEach(key => {
    const desc = (backup[key] = Object.getOwnPropertyDescriptor(obj, key))
    const newDesc: any = { ...desc }
    if (newDesc.get) {
      newDesc.get = () => (newProps as any)[key]
    } else {
      newDesc.value = (newProps as any)[key]
    }
    Object.defineProperty(obj, key, newDesc)
  })
  if ((obj as any).mockRestore) backup.mockRestore = Object.getOwnPropertyDescriptor(obj, 'mockRestore')
  return Object.defineProperty(obj, 'mockRestore', {
    value() {
      Object.keys(backup).forEach(key => {
        Object.defineProperty(obj, key, backup[key])
      })
      return obj
    },
    configurable: true,
  })
}

export const mockWriteStream = () => ({
  written: [] as string[],
  write(text: string) {
    this.written.push(text)
  },
  clear() {
    this.written = []
  },
})
