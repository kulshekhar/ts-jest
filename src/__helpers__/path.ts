import { ensureDirSync } from 'fs-extra'
import { tmpdir } from 'os'
import { join, resolve } from 'path'

export const ROOT = resolve(__dirname, '..', '..')

export function tempDir(ns: string): string {
  const dir = join(tmpdir(), 'ts-jest-unit-tests', ns, parseInt((Math.random() * 1e17) as any, 10).toString(36))
  ensureDirSync(dir)

  return dir
}
