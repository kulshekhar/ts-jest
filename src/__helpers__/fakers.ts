import { Config } from '@jest/types'
import { resolve } from 'path'

import { createCompilerInstance } from '../compiler/instance'
import { ConfigSet } from '../config/config-set'
import { BabelConfig, TsCompiler, TsJestConfig, TsJestGlobalOptions } from '../types'
import { ImportReasons } from '../util/messages'

export function filePath(relPath: string): string {
  return resolve(__dirname, '..', '..', relPath)
}

export const rootDir = filePath('')

export function tsJestConfig(options?: Partial<TsJestConfig>): TsJestConfig {
  return {
    isolatedModules: false,
    compiler: 'typescript',
    transformers: [],
    babelConfig: undefined,
    tsConfig: undefined,
    packageJson: undefined,
    stringifyContentPathRegex: undefined,
    diagnostics: { ignoreCodes: [], pretty: false, throws: true },
    ...options,
  }
}

function getJestConfig<T extends Config.ProjectConfig>(
  options?: Partial<Config.InitialOptions | Config.ProjectConfig>,
  tsJestOptions?: TsJestGlobalOptions,
): T {
  const res = {
    globals: {},
    ...options,
  } as any
  if (tsJestOptions) {
    res.globals['ts-jest'] = tsJestOptions
  }

  return res
}

export function babelConfig<T extends BabelConfig>(options?: BabelConfig): T {
  return {
    ...options,
    presets: [...(options && options.presets)],
    plugins: [...(options && options.plugins)],
  } as any
}

export function importReason(text = '[[BECAUSE]]'): ImportReasons {
  return text as any
}

export const defaultResolve = (path: string): string => `resolved:${path}`

export function createConfigSet({
  jestConfig,
  tsJestConfig,
  parentConfig,
  resolve = defaultResolve,
  ...others
}: {
  jestConfig?: Partial<Config.ProjectConfig>
  tsJestConfig?: TsJestGlobalOptions
  parentConfig?: TsJestGlobalOptions
  resolve?: ((path: string) => string) | null
  [key: string]: any
} = {}): ConfigSet {
  const defaultTestRegex = ['(/__tests__/.*|(\\\\.|/)(test|spec))\\\\.[jt]sx?$']
  const defaultTestMatch = ['**/__tests__/**/*.[jt]s?(x)', '**/?(*.)+(spec|test).[jt]s?(x)']
  jestConfig = {
    ...jestConfig,
    testMatch: jestConfig?.testMatch ? [...jestConfig.testMatch, ...defaultTestMatch] : defaultTestMatch,
    testRegex: jestConfig?.testRegex ? [...defaultTestRegex, ...jestConfig.testRegex] : defaultTestRegex,
  }
  const cs = new ConfigSet(getJestConfig(jestConfig, tsJestConfig), parentConfig)
  if (resolve) {
    cs.resolvePath = resolve
  }
  Object.keys(others).forEach((key) => {
    Object.defineProperty(cs, key, { value: others[key] })
  })

  return cs
}

// not really unit-testing here, but it's hard to mock all those values :-D
export function makeCompiler({
  jestConfig,
  tsJestConfig,
  parentConfig,
}: {
  jestConfig?: Partial<Config.ProjectConfig>
  tsJestConfig?: TsJestGlobalOptions
  parentConfig?: TsJestGlobalOptions
} = {}): TsCompiler {
  tsJestConfig = { ...tsJestConfig }
  tsJestConfig.diagnostics = {
    ...(tsJestConfig.diagnostics as any),
    pretty: false,
  }
  const cs = createConfigSet({ jestConfig, tsJestConfig, parentConfig, resolve: null })

  return createCompilerInstance(cs)
}
