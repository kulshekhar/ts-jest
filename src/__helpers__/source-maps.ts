// WARNING: this file is shared between e2e and unit tests
import bufferFrom = require('buffer-from')
import stableStringify = require('fast-json-stable-stringify')
import { realpathSync } from 'fs'
import { isAbsolute, relative } from 'path'
import { RawSourceMap } from 'source-map'

export function base64ToSourceMaps(base64: string): RawSourceMap {
  return JSON.parse(bufferFrom(base64, 'base64').toString('utf8'))
}

export function sourceMapsToBase64(sourceMaps: RawSourceMap): string {
  return bufferFrom(stableStringify(sourceMaps)).toString('base64')
}

export interface ParsedSourceWithMaps {
  sourceMaps?: RawSourceMap
  comment?: string
  source: string
}
export function parseSource(source: string): ParsedSourceWithMaps {
  const [comment, b64Maps]: [string, string | undefined] =
    // eslint-disable-next-line @typescript-eslint/prefer-regexp-exec
    (source.match(/[\n^]\/\/#\s*sourceMappingURL=data:application\/json;(?:charset=utf-8;)?base64,(\S+)\s*$/) as any) ||
    []
  if (b64Maps) {
    const map = base64ToSourceMaps(b64Maps)
    return {
      source: source.substr(0, -comment.length),
      comment: comment.trim(),
      sourceMaps: map,
    }
  }
  return { source }
}

// source-map module doesn't provide a sync method to extract the source-maps
export function extractSourceMaps(source: string): RawSourceMap | undefined {
  return parseSource(source).sourceMaps
}

export function relativisePaths(map: RawSourceMap, fromPath: string, newPrefix = ''): RawSourceMap {
  const res = { ...map }
  const from = realpathSync(fromPath)
  const remap = (path: string): string =>
    (isAbsolute(path) ? `${newPrefix}${relative(from, realpathSync(path))}` : path).replace(/\\/g, '/')
  if (res.sourceRoot) res.sourceRoot = remap(res.sourceRoot)
  if (res.sources) res.sources = res.sources.map(remap)
  if (res.file) res.file = remap(res.file)
  return res
}

export function rewriteSourceMaps(source: string, sourceMapsTransformer: (maps: RawSourceMap) => RawSourceMap): string {
  return source.replace(
    /([\n^]\/\/#\s*sourceMappingURL=data:application\/json;(?:charset=utf-8;)?base64,)(\S+)(\s*)$/,
    (_, before, base64, after) => {
      let map = base64ToSourceMaps(base64)
      map = sourceMapsTransformer(map)
      return `${before}${sourceMapsToBase64(map)}${after}`
    },
  )
}
