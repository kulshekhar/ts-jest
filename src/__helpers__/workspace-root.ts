import fs from 'node:fs'
import path from 'node:path'

function workspaceRootInner(dir: string, candidateRoot: string) {
  if (path.dirname(dir) === dir) return candidateRoot

  const matchablePaths = [path.join(dir, '.ts-jest-digest'), path.join(dir, '.github'), path.join(dir, 'renovate.json')]
  if (matchablePaths.some((x) => fs.existsSync(x))) {
    return dir
  }

  return workspaceRootInner(path.dirname(dir), candidateRoot)
}

export const workspaceRoot = workspaceRootInner(process.cwd(), process.cwd())
