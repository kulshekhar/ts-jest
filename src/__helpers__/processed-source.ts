import { isAbsolute, relative } from 'path'
import { RawSourceMap } from 'source-map'

import { ROOT } from './path'
import { ParsedSourceWithMaps, parseSource, relativisePaths, rewriteSourceMaps } from './source-maps'

export default class ProcessedSource {
  readonly filename: string

  constructor(readonly output: string, filename: string, readonly cwd: string = ROOT) {
    if (isAbsolute(filename)) filename = relative(this.cwd, filename)
    this.filename = filename
  }
  get outputCode(): string {
    return this.output
  }
  get parsedSource(): ParsedSourceWithMaps {
    return parseSource(this.output)
  }
  get outputCodeWithoutMaps(): string {
    return this.parsedSource.source
  }
  get outputSourceMaps(): RawSourceMap | undefined {
    return this.parsedSource.sourceMaps
  }
  get normalizedOutputCode(): string {
    return rewriteSourceMaps(this.output, this.sourceMapsNormalizer)
  }
  get normalizedOutputSourceMaps(): RawSourceMap | undefined {
    const maps = this.outputSourceMaps
    if (maps) return this.sourceMapsNormalizer(maps)

    return undefined
  }
  get sourceMapsNormalizer() {
    return (maps: RawSourceMap): RawSourceMap => relativisePaths(maps, this.cwd, '<cwd>/')
  }
}
