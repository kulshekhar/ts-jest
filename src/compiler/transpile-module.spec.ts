import { LogLevels } from 'bs-logger'
import { removeSync, writeFileSync } from 'fs-extra'

import { makeCompiler } from '../__helpers__/fakers'
import { logTargetMock } from '../__helpers__/mocks'
import ProcessedSource from '../__helpers__/processed-source'

const logTarget = logTargetMock()

describe('transpile module with isolatedModule: true', () => {
  const baseTsJestConfig = {
    isolatedModules: true,
  }

  beforeEach(() => {
    logTarget.clear()
  })

  it('should compile using transpileModule and not use cache', () => {
    const compiler = makeCompiler({ tsJestConfig: { ...baseTsJestConfig, tsConfig: false } }),
      spy = jest.spyOn(require('typescript'), 'transpileModule')

    logTarget.clear()
    const compiled = compiler.compile('export default 42', __filename)

    expect(new ProcessedSource(compiled, __filename)).toMatchSnapshot()
    expect(spy).toHaveBeenCalled()
    expect(logTarget.filteredLines(LogLevels.debug, Infinity)).toMatchInlineSnapshot(`
      Array [
        "[level:20] readThrough(): no cache
      ",
        "[level:20] getOutput(): compiling as isolated module
      ",
        "[level:20] visitSourceFileNode(): hoisting
      ",
      ]
    `)

    spy.mockRestore()
  })

  it('should compile js file for allowJs true', () => {
    const fileName = `${__filename}.test.js`,
      compiler = makeCompiler({
        tsJestConfig: { ...baseTsJestConfig, tsConfig: { allowJs: true, outDir: '$$ts-jest$$' } },
      }),
      source = 'export default 42'

    writeFileSync(fileName, source, 'utf8')
    const compiled = compiler.compile(source, fileName)

    expect(new ProcessedSource(compiled, fileName)).toMatchSnapshot()

    removeSync(fileName)
  })

  it('should compile tsx file for jsx preserve', () => {
    const fileName = `foo.tsx`,
      compiler = makeCompiler({
        tsJestConfig: {
          ...baseTsJestConfig,
          tsConfig: {
            jsx: 'preserve' as any,
          },
        },
      }),
      source = `
        const App = () => {
          return <>Test</>
        }
      `

    writeFileSync(fileName, source, 'utf8')
    const compiled = compiler.compile(source, fileName)

    expect(new ProcessedSource(compiled, fileName)).toMatchSnapshot()

    removeSync(fileName)
  })

  it('should compile tsx file for other jsx options', () => {
    const fileName = `foo.tsx`,
      compiler = makeCompiler({
        tsJestConfig: {
          ...baseTsJestConfig,
          tsConfig: {
            jsx: 'react' as any,
          },
        },
      }),
      source = `
        const App = () => {
          return <>Test</>
        }
      `

    writeFileSync(fileName, source, 'utf8')
    const compiled = compiler.compile(source, fileName)

    expect(new ProcessedSource(compiled, fileName)).toMatchSnapshot()

    removeSync(fileName)
  })

  it('should have correct source maps', () => {
    const compiler = makeCompiler({ tsJestConfig: { ...baseTsJestConfig, tsConfig: false } }),
      source = 'const f = (v: number) => v\nconst t: number = f(5)'

    const compiled = compiler.compile(source, __filename)

    expect(new ProcessedSource(compiled, __filename).outputSourceMaps).toMatchObject({
      file: __filename,
      sources: [__filename],
      sourcesContent: [source],
    })
  })

  it('should not report diagnostics related to typings', () => {
    const compiler = makeCompiler({ tsJestConfig: { ...baseTsJestConfig, tsConfig: false } })

    expect(() =>
      compiler.compile(
        `
const f = (v: number) => v
const t: string = f(5)
const v: boolean = t
`,
        'foo.ts',
      ),
    ).not.toThrowError()
  })

  it('should report diagnostics related to codes with pathRegex config is undefined', () => {
    const compiler = makeCompiler({ tsJestConfig: { ...baseTsJestConfig, tsConfig: false } })

    expect(() =>
      compiler.compile(
        `
const f = (v: number) = v
const t: string = f(5)
`,
        'foo.ts',
      ),
    ).toThrowErrorMatchingSnapshot()
  })

  it('should report diagnostics related to codes with pathRegex config matches file name', () => {
    const compiler = makeCompiler({
      tsJestConfig: { ...baseTsJestConfig, tsConfig: false, diagnostics: { pathRegex: 'foo.ts' } },
    })

    expect(() =>
      compiler.compile(
        `
const f = (v: number) = v
const t: string = f(5)
`,
        'foo.ts',
      ),
    ).toThrowErrorMatchingSnapshot()
  })

  it('should not report diagnostics related to codes with pathRegex config does not match file name', () => {
    const compiler = makeCompiler({
      tsJestConfig: { ...baseTsJestConfig, tsConfig: false, diagnostics: { pathRegex: 'bar.ts' } },
    })

    expect(() =>
      compiler.compile(
        `
const f = (v: number) = v
const t: string = f(5)
`,
        'foo.ts',
      ),
    ).not.toThrowError()
  })
})
