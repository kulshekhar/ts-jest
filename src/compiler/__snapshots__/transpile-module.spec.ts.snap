// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transpile module with isolatedModule: true should compile js file for allowJs true 1`] = `
  ===[ FILE: src/compiler/transpile-module.spec.ts.test.js ]======================
  "use strict";
  Object.defineProperty(exports, "__esModule", { value: true });
  exports.default = 42;
  //# sourceMappingURL=data:application/json;charset=utf-8;base64,eyJmaWxlIjoiPGN3ZD4vc3JjL2NvbXBpbGVyL3RyYW5zcGlsZS1tb2R1bGUuc3BlYy50cy50ZXN0LmpzIiwibWFwcGluZ3MiOiI7O0FBQUEsa0JBQWUsRUFBRSxDQUFBIiwibmFtZXMiOltdLCJzb3VyY2VzIjpbIjxjd2Q+L3NyYy9jb21waWxlci90cmFuc3BpbGUtbW9kdWxlLnNwZWMudHMudGVzdC5qcyJdLCJzb3VyY2VzQ29udGVudCI6WyJleHBvcnQgZGVmYXVsdCA0MiJdLCJ2ZXJzaW9uIjozfQ==
  ===[ INLINE SOURCE MAPS ]=======================================================
  file: <cwd>/src/compiler/transpile-module.spec.ts.test.js
  mappings: ';;AAAA,kBAAe,EAAE,CAAA'
  names: []
  sources:
    - <cwd>/src/compiler/transpile-module.spec.ts.test.js
  sourcesContent:
    - export default 42
  version: 3
  ================================================================================
`;

exports[`transpile module with isolatedModule: true should compile using transpileModule and not use cache 1`] = `
  ===[ FILE: src/compiler/transpile-module.spec.ts ]==============================
  "use strict";
  Object.defineProperty(exports, "__esModule", { value: true });
  exports.default = 42;
  //# sourceMappingURL=data:application/json;charset=utf-8;base64,eyJmaWxlIjoiPGN3ZD4vc3JjL2NvbXBpbGVyL3RyYW5zcGlsZS1tb2R1bGUuc3BlYy50cyIsIm1hcHBpbmdzIjoiOztBQUFBLGtCQUFlLEVBQUUsQ0FBQSIsIm5hbWVzIjpbXSwic291cmNlcyI6WyI8Y3dkPi9zcmMvY29tcGlsZXIvdHJhbnNwaWxlLW1vZHVsZS5zcGVjLnRzIl0sInNvdXJjZXNDb250ZW50IjpbImV4cG9ydCBkZWZhdWx0IDQyIl0sInZlcnNpb24iOjN9
  ===[ INLINE SOURCE MAPS ]=======================================================
  file: <cwd>/src/compiler/transpile-module.spec.ts
  mappings: ';;AAAA,kBAAe,EAAE,CAAA'
  names: []
  sources:
    - <cwd>/src/compiler/transpile-module.spec.ts
  sourcesContent:
    - export default 42
  version: 3
  ================================================================================
`;

exports[`transpile module with isolatedModule: true should report diagnostics related to codes with pathRegex config is undefined 1`] = `
"TypeScript diagnostics (customize using \`[jest-config].globals.ts-jest.diagnostics\` option):
foo.ts(2,23): error TS1005: '=>' expected."
`;

exports[`transpile module with isolatedModule: true should report diagnostics related to codes with pathRegex config matches file name 1`] = `
"TypeScript diagnostics (customize using \`[jest-config].globals.ts-jest.diagnostics\` option):
foo.ts(2,23): error TS1005: '=>' expected."
`;
